money-core/src/main/scala/com/comcast/money/core/ConfigurableTypeFactory.scala

```diff
@@ -0,0 +1,124 @@
+
+import java.lang.reflect.Modifier
+import scala.reflect.ClassTag
+import scala.util.Try
+
+/**
+ * Helper trait used to create plugin factories that can create instances of a plugin
+ * from a common configuration format.
+ *
+ * @tparam T the trait of the plugin
+ */
+trait ConfigurableTypeFactory[T <: AnyRef] {
+  private val TYPE_KEY: String = "type"
+  private val CLASS_KEY: String = "class"
+  private val APPLY_NAME: String = "apply"
+  private val MODULE_SUFFIX: String = "$"
+  private val MODULE_FIELD_NAME: String = "MODULE$"
+
+  protected val tag: ClassTag[T]
+  protected val logger: Logger = LoggerFactory.getLogger(getClass)
+  protected val knownTypes: PartialFunction[String, Config => T] = Map.empty
+  protected val defaultValue: Option[T] = None
+
+  def create(config: Config): Option[T] =
+    findKnownType(config)
+      .orElse(createInstance(config))
+      .map(_(config))
+      .orElse(defaultValue)
+
+  protected def findConfigValue(config: Config, key: String): Option[String] =
+    Option(config)
+      .filter(_.hasPath(key))
+      .map(_.getString(key))
+      .filterNot(_.isEmpty)
+
+  protected def findKnownType(config: Config): Option[Config => T] =
+    findConfigValue(config, TYPE_KEY)
+      .flatMap(findKnownType)
+
+  protected def findKnownType(typeName: String): Option[Config => T] =
+    knownTypes.lift(typeName)
+      .orElse({
+        logger.warn(s"Could not find ${tag.runtimeClass.getSimpleName} type: '$typeName'")
+        None
+      })
+
+  protected def createInstance(config: Config): Option[Config => T] =
+    findConfigValue(config, CLASS_KEY)
+      .flatMap(findConfigurableTypeFactory)
+
+  protected def findConfigurableTypeFactory(className: String): Option[Config => T] =
+    findClass(className)
+      .filter(tag.runtimeClass.isAssignableFrom)
+      .flatMap({
+        classInstance =>
+          findFactoryMethod(classInstance)
+            .orElse(findConfigConstructor(classInstance))
+            .orElse(findDefaultConstructor(classInstance))
+      })
+      .orElse(findModule(className)
+        .flatMap({
+          moduleInstance => findStaticInstance(moduleInstance)
+        }))
+      .orElse({
+        logger.warn(s"Could not create ${tag.runtimeClass.getSimpleName} class: '$className'")
+        None
+      })
+
+  private def findClass(className: String): Option[Class[_]] =
+    Try(Class.forName(className)).toOption
+
+  private def findModule(moduleName: String): Option[Class[_]] =
+    findClass(moduleName + MODULE_SUFFIX)
+
+  private def findFactoryMethod(cls: Class[_]): Option[Config => T] =
+    cls.getMethods.find {
+      method =>
+        Modifier.isStatic(method.getModifiers) &&
+          method.getParameterCount == 1 &&
+          method.getParameterTypes()(0) == classOf[Config] &&
+          method.getReturnType == cls &&
+          method.getName == APPLY_NAME
+    } map { method => config: Config => method.invoke(null, config).asInstanceOf[T] }
```

Comment:
infix with alphabetic methods is deprecated in scala 3
```suggestion
    }
    .map { method => config: Config => method.invoke(null, config).asInstanceOf[T] }
```